### PR TITLE
remove bad ES rechits from PF clustering

### PR DESCRIPTION
**Pull request by cmkuo**
remove ES rechits with bad status flag from PF ES clustering

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@cmkuo please quantify the significance of this change for physics dataset (re)processing.

**Comment by cmkuo** [comment]
This fix should remove bad ES hits from clustering so the effect is we will get less energy in PF ES clustering comparing to what we have in prompt reco data currently. This fix was tested within ECAL DPG and shown in ECAL calibration workshop.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by davidlange6** [comment]
the block we had for the 1003 workflow is gone from cern. rerunning won’t help

**Comment by lgray** [comment]
@cmkuo could you post a few plots showing what changes (or the link to the ECAL workshop slides) here?

Likewise, talking with Fabrice this is known improve data/MC agreement for the shape of the preshower energy. However, to get the normalization correct it sounded like some conditions would need to be added to the MC to approximate the dead channel list. What's the plan for this?

**Comment by lgray** [comment]
@davidlange6 Noted, won't rerun them.

**Comment by lgray** [comment]
Nice, looks like a bigger infrastructure error has taken its place.

**Comment by lgray** [comment]
@fcouderc FYI

**Comment by davidlange6** [comment]
no, just a code bug:(

**Comment by lgray** [comment]
ouch :-(

**Comment by smuzaffar** [comment]
@cmsbuild, please test
CMSSW_7_6_X_2015-10-16-2300 is marked as bad release now

**Comment by slava77** [comment]
@cmkuo 
If I understand the context correctly, this is a small improvement over the validated state of the PF code that was available for a couple of years already.
So, if it produces noticeable differences, it's too late for 76X release.

**Comment by lgray** [comment]
@slava77 It's (a big improvement towards) fixing a fairly clear data-MC discrepancy for the ES. The rest would likely come from updating the dead channel lists in the MC.

**Comment by slava77** [comment]
On 10/18/15 10:58 AM, Lindsey Gray wrote:
> @slava77 <https://github.com/slava77> It's (a big improvement towards)
> fixing a fairly clear data-MC discrepancy for the ES. The rest would
> likely come from updating the dead channel lists in the MC.

Some clear plots with gains are needed  (slides in a meeting could work)
and PC should be made aware that we may have to delay the end-of-the-year
processing schedule and confirm that this feature is needed.

To me, this PR is a new feature.

**Comment by fcouderc** [comment]
Here are a few plots showing the improvements when adding the cleaning.
Before cleaning for ES+
image

After cleaning:
image


So the cleaning removes the excess of low energy ES clusters.

**Comment by fcouderc** [comment]
FYI @emanueledimarco

**Comment by lgray** [comment]
@fcouderc your plots did not attach

**Comment by slava77** [comment]
@fcouderc please re-post the plots.
They don't show up correctly.
Also, details of how they were made would be nice.
Thanks.

**Comment by fcouderc** [comment]
tyring again to upload the plots
Before cleaning:
screen shot 2015-10-19 at 16 09 10

After cleaning:
screen shot 2015-10-19 at 16 09 46


So the cleaning removes the excess of low energy ES clusters.

**Comment by fcouderc** [comment]
Plots are showing the energy deposited in ES+ from Z->ee electrons requiring that these electrons are not in bad regions of the ES (known hardware failure) for three different data taking periods:
1. runNumber >= 254227 && runNumber <= 255031
2. runNumber >= 256605 && runNumber <= 257426
3. runNumber >= 257421
These 3 different running periods have different hardware ES problems 

The ES energy in data was also increased by 10% but this is a complete other issue.

One sees that the excess of low energy clusters are basically moved down to zero energy cluster which indicates that these low energy clusters are actually due to bad channels.


Same kind of plots were done for ES- showing similar conclusions

**Comment by slava77** [comment]
@cmsbuild please test
I think the file in 1003.0 is now available again

**Comment by slava77** [comment]
this is formally a "-1"

for https://github.com/cms-sw/cmssw/pull/11858	 b3b69e0
- changes in the code are as described, but they are apparently not enough
- tested locally in CMSSW_7_6_X_2015-10-19-1100 with higher statistics tests to see distributions better than just in jenkins comparisons
    - the primary discrepancy is in the particleFlowClusterPS distributions, which change rather consistently both in data and in MC
    - the  physics objects downstream of particleFlowClusterPS are not affected substantially (the samples do not target problematic data taking periods and do not have a high enough statistics in signal egamma objects): there is some noticeable variation in mva_e_pi for ged electrons but not enough stats to see it's significant
    - reduction in the particleFlowClusterPS counts turns into a huge blow up in the particleFlowRecHitPS_Cleaned collection; in 25ns data AOD files increase in size by a factor of 3!
    - smaller size of passing PS hits leads to faster clustering and PF block reconstruction: 
        - in 25202 (ttbar with PU35@25ns) particleFlowClusterPS is x3 faster, particleFlowBlock is x2 faster, particleFlowClusterECAL is almost x2 faster; the total reduction here is about 0.4 s/evt or about 2% of the reco sequence
        - the effect is smaller at lower PU

Here are some plots 
in 134.802 workflow (2015C DoubleEG): the reduction in the count by ~x2 is about the same in MC as well; in 50 ns run2 and run1 data as well as in MC the reduction in the number of clusters is about x1.5
all_sign618vsorig_rundoubleeg2015cwf134p802c_log10recopfclusters_particleflowclusterps__rereco_obj_energy

in 202.0 workflow (ttbar run1 with 50ns ): just one of the more changing variables (single particle guns are fairly clean/unchaged) maybe egamma objects in jets are impacted somewhat
all_sign618vsorig_ttbarpuwf202p0c_recogsfelectrons_gedgsfelectrons__reco_obj_mvaoutput_mva_e_pi

Are there any plots suggesting improvements at the high level objects (ged electrons or photons: efficiencies or resolutions)?
Plots posted by @fcouderc  earlier https://github.com/cms-sw/cmssw/pull/11858#issuecomment-149225428 are for a rather low-level targeted variable.

The issue with particleFlowRecHitPS_Cleaned collection picking up all of the cleaned up hits has to be addressed first, before it can move on even in 80X: I don't think we can keep an extra 400kB even in RECO. Either write the more "cleaner" cleaned collection or drop it completely (isn't it needed, actually?).

**Comment by cmkuo** [comment]
@slava77 a threshold is added to the "cleaned" ES rechits. this should make the cleaned collection smaller

File: RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h [hunk 1]
```diff
@@ -425,9 +425,74 @@ class PFRecHitQTestECAL : public PFRecHitQTestBase {
...
+    topologicalCleaning_ = iConfig.getParameter<bool>("topologicalCleaning");
+  }
+
+  void beginEvent(const edm::Event& event,const edm::EventSetup& iSetup) {
+  }
+
+  bool test(reco::PFRecHit& hit,const EcalRecHit& rh,bool& clean){
+
+    if ( rh.energy() < thresholdCleaning_ ) {
+      clean=true;
```
Comment (lgray) [inline RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h]: Hi Ming, this should be:
```
clean = false;
return false;
```

**Comment by cmkuo** [comment]
@lgray thanks for pointing this out. it's fixed.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmkuo 
may I repeat my question earlier in this PR:
"""
Are there any plots suggesting improvements at the high level objects (ged electrons or photons: efficiencies or resolutions)?
Plots posted by @fcouderc earlier #11858 (comment) are for a rather low-level targeted variable.
"""

**Comment by lgray** [comment]
@slava77 It should improve data/MC agreement for electrons and photons in the endcap. Otherwise there is no particular 'improvement' expected. More that things are simply done correctly and consistently in data and MC.

**Comment by emanueledimarco** [comment]
@fcouderc I have a question on the plots: do they include also the ad-hoc shift of the bulk energy of 10% (which we don't know why we should apply it, since the calibrations seem to go in the opposite direction) ?

**Comment by slava77** [comment]
On 10/23/15 5:28 AM, Lindsey Gray wrote:
> @slava77 <https://github.com/slava77> It should improve data/MC
> agreement for electrons and photons in the endcap. Otherwise there is no
> particular 'improvement' expected.

Is there a plot for this?
Just to understand the significance for 76X
(this PR is a new feature and it is added at a fairly low level;
so, a clear reason for an exception to include it now should be seen)

**Comment by lgray** [comment]
@cmkuo @fcouderc Could you guys make plots of endcap photon ID BDT scores in data and MC? As well as E_RECO/E_True distributions for electrons/photons in the endcap?

This would make a fine demonstration of the impact.

**Comment by emanueledimarco** [comment]
@cmkuo @fcouderc and please, do the plot E_RECO/E_True as @lgray suggests w/o the 10% rescaling which is unjustified as far as we know

**Comment by lgray** [comment]
@emanueledimarco If they just plot the stock E_reco/E_true of objects then there's no introduction of 10% offset. I'd prefer that this be done too.

**Comment by slava77** [comment]
Hi
Should we expect some plots by tomorrow?

**Comment by cmkuo** [comment]
@lgray I can do the plots. However, I need to have re-reconstructed double EG dataset with removing bad ES rechits. @emanueledimarco Where can I find the dataset produced by ECAL DPG ?

**Comment by emanueledimarco** [comment]
@cmkuo Files of the rereco are on EOS here:
    /eos/cms/store/group/dpg_ecal/alca_ecalcalib/ecalelf/alcarereco/13TeV/preshowerClustering_v1/

including @shervin86, he also has validation plots of data/MC with the rereco

**Comment by slava77** [comment]
this is still a "-1"
for purely technical reasons:
based on the latest updates in  https://github.com/cms-sw/cmssw/pull/11858	 36e5650
in 134.802:
the recoPFRecHits_particleFlowRecHitPS_Cleaned is now only 350 kB/event instead of 359 kB.
This should be reduced by about a factor of 1000 before we can let this go in the release.

**Comment by cmkuo** [comment]
@slava77 do you know the size of recoPFRecHits_particleFlowRecHitPS_Cleaned before the fix was added ?

**Comment by slava77** [comment]
@cmkuo 
it was 15 bytes/event

**Comment by cmkuo** [comment]
after taking to Lindsey, it makes more sense to put the cleaning threshold to the seeding threshold. the change was just committed. can you please test it ?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
After the last update (in the same 134.802), the per-event size of recoPFRecHits_particleFlowRecHitPS_Cleaned is now 95kB. This is too large for AOD.

You can tune this up better without me having to rerun and test: ``` edmEventSize -v fileName.root ```  will tell you the event size.

Based on what I understand the need of this collection is (saving some fairly significant or rare bad hits), I'd expect a reasonable point is around 1kB/event (less is better).

**Comment by emanueledimarco** [comment]
@cmkuo do we really need the cleaned collection at all?

**Comment by slava77** [comment]
Since this was discussed in RECO meeting:
https://cloud.githubusercontent.com/assets/4676718/10621215/edfab852-7744-11e5-9c88-b38f08184a49.png
the plot from 2015C in data (see above; reproduce with ``` runTheMatrix.py -l 134.802 -i all ``` )
shows that a very large fraction of clusters disappear after the change; this has to be a result of reduction in the number of input hits.
Why is this happening?

Does this mean there is a large fraction (~50%) of hits that are actually bad?
If this is real, I think we should not even make these hits at this fraction of badness.
It makes sense to keep bad hits in RECO outputs only if they are a small fraction of the total (few % or less).
ES local reconstruction should not make such a large rate of useless output.

**Comment by slava77** [comment]
here is an example of the plot in simulation single electron 35 GeV (workflow 1317.0 in runTheMatrix.py)

all_sign618vsorig_singleelectron13pt35wf1317p0c_log10recopfclusters_particleflowclusterps__reco_obj_energy

clearly, some noise is reconstructed by local reco and is being cleaned up even in simulation.
This kind of hits are some generic noise that don't indicate any detector problem. I'd expect it should be possible to not make them at local reconstruction step.

**Comment by emanueledimarco** [comment]
@diguida @cmkuo @shervin86 (moving a private discussion here)
I was expecting that, based on the plots shown by Fabrice earlier, the cleaning removes the 
bad rechits which populated the low energy shoulder of the distribution, so the change should go in the opposite direction (increasing the energy). Isn't it?

**Comment by fcouderc** [comment]
@emanueledimarco @diguida @cmkuo @shervin86 @slava77

Is the plot we are looking at here really the log(energy) ? So, 99% of all these cluster have an energy below 0.01 GeV, and most of them (like 90%), energies below 0.001 GeV?

Emanuele, the clusters we are looking at, even in the so called low energy shoulder have energies higher than 0.1 GeV and do no even appear in this plot, you can not draw any conclusion from real physics clusters and these super low energy clusters.

I think we should look at the effect of the cleaning on the clusters that are not in this plot, the ones, for which you do not need to use the log scale.

Do we even need to keep ES clusters below 10ˆ-3 GeV? Do they really change anything to any physics?

**Comment by lgray** [comment]
@fcouderc Considering we seed clusters at 1e-4, yes. Remember, this is energy not corrected for absorber thickness (just Si readout).

**Comment by fcouderc** [comment]
@lgray, ah ok, so what's the correction to apply in the end on top of the energy here? Is it roughly gamma (~200)?

**Comment by lgray** [comment]
@fcouderc Around that... but it's probably better just to look at raw signal spectra vs noise...

**Comment by slava77** [comment]
Any update here yet?

**Comment by cmkuo** [comment]
@slava77 we decided to keep the collection but remove all bad ES rechits

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Do we have a sign-off from the Physics Coordination for this change to be in 76X?

**Comment by slava77** [comment]
(760 validation deadline is later this week on Nov 6)

**Comment by cmkuo** [comment]
@emanueledimarco @lgray @fcouderc @matteosan1 
If we would like this removal to go int 76X, we need the green light from physics coordination.
can you please help if you agree this should go into 76X ?

**Comment by shervin86** [comment]
Dear all,
in slide 4 of https://indico.cern.ch/event/402605/contribution/4/attachments/1182594/1713434/DPG-ES_v1.pdf
you can find the effect of the topological cleaning in this PR on electrons from Zee decay.
In the right plot you can see a comparison of the reference (prompt, no cleaning), ES cleaning and MC.

One caveat is that the validation has been done with only the topological cut and not with the energy cut in line 450

**Comment by emanueledimarco** [comment]
@slava77 are we still on time to include this in 76X? It would be beneficial not only for just ES energy, but also to the total energy of e/g because we have ES energy in the regression trained with MC, so data/MC discrepancies could be amplified by the regression in the total energy. Thanks (I know it's late).

**Comment by davidlange6** [comment]
No, not without a request to delay things from PC  (deadline was ~a month ago).

File: RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitPS_cfi.py [hunk 2]
```diff
@@ -5,17 +5,22 @@
...
+            name = cms.string("PFPSRecHitCreator"),
+            src  = cms.InputTag("ecalPreshowerRecHit","EcalRecHitsES"),
+            qualityTests = cms.VPSet(
+                cms.PSet(
+                    name = cms.string("PFRecHitQTestThreshold"),
+                    threshold = cms.double(7e-6)
+                    ),
+                cms.PSet(
+                    name = cms.string("PFRecHitQTestES"),
+                    cleaningThreshold = cms.double(1400.),
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitPS_cfi.py]: is it the right threshold to change?

**Comment by slava77** [comment]
It looks like the last update killed all the ES rechits for the PF purposes:
all_oldvsnew_zmm13tevwf1330p0c_log10recopfclusters_particleflowclusterps__reco_obj_energy
this is from jenkins workflow 1330 comparison
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-11-03-2300+11858/11131/validateJR/all_OldVSNew_ZMM13TeVwf1330p0/all_OldVSNew_ZMM13TeVwf1330p0c_log10recoPFClusters_particleFlowClusterPS__RECO_obj_energy.png

**Comment by slava77** [comment]
@cmkuo 
please clarify if we should expect an update soon

**Comment by slava77** [comment]
-1
MC DR 76X campaign has essentially started.
This change was agreed to go to 80X.
We may still exchange plots in this thread, but the target PR for integration becomes just #11873.

**Comment by davidlange6** [comment]
I am cleaning up the 76x queue aside from things for analysis workflows. I'm closing this pull request, please make sure the PR is in 80x. Thanks!

**Pull request closed without merging**